src/hooks/stopwatch.ts

```diff
@@ -0,0 +1,63 @@
+import * as React from "react";
+
+type StopwatchState = {
+  time: number;
+  stopped: boolean;
+};
+type StopwatchActions = "TICK" | "STOP" | "START";
+type StopwatchReducer = (
+  state: StopwatchState,
+  action: StopwatchActions,
+) => StopwatchState;
+
+export type VoidFunction = () => void;
+export type StopwatchHook = [number, VoidFunction, VoidFunction];
+
+export function useStopwatch(startTime: number): StopwatchHook {
+  const tickSize = 10;
+  const initState = {
+    time: startTime,
+    stopped: true,
+  };
+  const [state, dispatch] = React.useReducer<StopwatchReducer>(
+    (state, action) => {
+      switch (action) {
+        case "TICK":
+          return {
+            ...state,
+            time: state.time + tickSize,
+          };
+        case "STOP":
+          return {
+            ...state,
+            stopped: true,
+          };
+        case "START":
+          return {
+            ...state,
+            stopped: false,
+          };
+        default:
+          return state;
+      }
+    },
+    initState,
+  );
+
+  React.useEffect(() => {
+    if (!state.stopped) {
+      const id = setTimeout(() => dispatch("TICK"), tickSize);
```

Comment:
so this technically works, but I don't think using `setTimeout` as-is is a reliable timer unless things has changed 🤔. https://stackoverflow.com/a/29972322

But as a proof of concept with hooks this is cool, I haven't worked with hooks yet

Reply:
switched to using `performance.now()`. From my understanding that's the "better" api to use for measuring time.